DOCS: Revise typescript docs to be user-centric (#1940)
StackExchange/dnscontrol: docs/typescript.md

---
layout: default
title: Using TypeScript with DNSControl
---

# Using TypeScript with DNSControl (experimental)

## What is this?

Would you like your editor to support auto-completion and other advanced IDE
features when editing dnsconfig.js?  Yes you can!

While DNSControl does not support TypeScript syntax in `dnsconfig.js`, you can
still use TypeScript’s features in editors which support it.

If you’re using VS Code (or another editor that supports TypeScript), you
should now be able to see the type information in your `dnsconfig.js` file as
you type. Hover over record names to read their documentation without having to
open the website!

## How to activate

To set up TypeScript support in Visual Studio Code, follow these steps:

1. Run this command to generate the file `types-dnscontrol.d.ts`.

```bash
dnscontrol write-types
```

This file has all the information your editor or IDE needs.  It must be in the same directory as the `dnsconfig.js` file you are editing.

> NOTE: Re-run the `dnscontrol write-types` command any time you upgrade
> dnscontrol. Because it is generated from the command, it will always be correct
> for the version of DNSControl you are using.

Reviewer: This is a little unclear in my eyes (but I’m not sure if my suggestion makes it any better). I wanted to clarify that it’s downloaded from the Internet, and that it matches the current version of DNSControl rather than just grabbing the latest version.

```suggestion
NOTE: Re-run the `dnscontrol write-types` command any time you upgrade
dnscontrol. The file will always match the version of DNSControl that is
currently running, regardless of what the latest version is.
```

Author: Oh! You are correct!  I forgot that the code downloads the latest version.  Would it be better if it downloaded the version that is appropriate for this version of dnscontrol?

There's an easy way to implement that.  Go can embed the file into the binary. The write-types command would write the contents to the file instead of downloading it.

Reviewer: > Oh! You are correct! I forgot that the code downloads the latest version. Would it be better if it downloaded the version that is appropriate for this version of dnscontrol?

It should be downloading the version corresponding to the commit SHA used to build DNSControl. I didn’t embed it partially because I don’t really know how to do so but also because it’s around 100kB uncompressed and I wouldn’t want to bloat the binary too much (although I guess it isn’t too bad compared to the ~40-50MB total binary size)

Author: I'm ok with 100kB being added.  This way people that build and compile their own binaries will be supported.

Reviewer: Will open a PR to do that then.

> 2. Tell your editor

Reviewer: This is only necessary to get type checking (i.e. red squiggly underlines when you misuse APIs), autocomplete will work without `@ts-check`.

Author: I'll add a note.

Add this comment to the top of your `dnsconfig.js` file:

```
// @ts-check
```

That should be all you need to do!

If your edit requires extra steps, please [file a bug](https://github.com/StackExchange/dnscontrol/issues) and we'll update this page.

### Bugs?

**Bugs?**  Not all features of DNSControl work perfectly at the moment. Please report bugs and feature requests on https://github.com/StackExchange/dnscontrol/issues

**This is experimental.** This feature is currently experimental. We might change the installation instructions as we find better ways to enable this.

## Known bugs

## Bug: `CLI_DEFAULTS` not implemented

Bug: Values passed to `CLI_DEFAULTS` (and the corresponding `-v` command-line option) don’t show up as global variables

> Workaround: create a new `.d.ts` file in the same folder as your `dnsconfig.js` file. In that file, add the following line for each variable you want to use. 

Reviewer: ```suggestion
Workaround: create a new `.d.ts` file in the same folder as your `dnsconfig.js` file. In that file, add the following line for each variable you want to use (replacing `VARIABLE_NAME` with the name of the variable. 
```

```

> <code>declare const _variableName_: string;</code>

Reviewer: ```suggestion
declare const VARIABLE_NAME: string;
```

---
layout: default
title: Using TypeScript with DNSControl
---

# Using TypeScript with DNSControl (experimental)

## What is this?

Would you like your editor to support auto-completion and other advanced IDE

> features when editing dnsconfig.js?  Yes you can!

Reviewer 2: ```suggestion
features when editing `dnsconfig.js`? Yes you can!
```

While DNSControl does not support TypeScript syntax in `dnsconfig.js`, you can
still use TypeScript’s features in editors which support it.

> If you’re using VS Code (or another editor that supports TypeScript), you

Reviewer 2: ```suggestion
If you’re using Visual Studio Code (or another editor that supports TypeScript), you
```

Reviewer 2: So the same spelling is used as https://github.com/StackExchange/dnscontrol/pull/1940/files#diff-ae03a9f2eb638a25c1d8d7ba853c78220f1ed44a19ad97a8d780cbc0c00a4260R23

should now be able to see the type information in your `dnsconfig.js` file as
you type. Hover over record names to read their documentation without having to

> open the website!

Reviewer 2: ```suggestion
open the documentation website!
```

> ## How to activate

Reviewer 2: ```suggestion
## How to activate auto-completion
```

> To set up TypeScript support in Visual Studio Code, follow these steps:

Reviewer 2: How about other IDE? Would you like to describe them all? For example the [JetBrains PHStorm documentation](https://www.jetbrains.com/help/phpstorm/typescript-support.html) is very complete and makes for me no sense to copy/paste all there content. What do you think?

Reviewer: These steps should be the same for all editors.

> 1. Run this command to generate the file `types-dnscontrol.d.ts`.

Reviewer 2: I have not tested the TypeScript setup before, I am not getting a populated `types-dnscontrol.d.ts` file at the moment.

```shell
brew upgrade --greedy
```

```text
==> Upgrading dnscontrol
  3.23.0 -> 3.24.0

==> Pouring dnscontrol--3.24.0.ventura.bottle.tar.gz
/usr/local/Cellar/dnscontrol/3.24.0: 5 files, 35.3MB
==> Running `brew cleanup dnscontrol`...
Removing: /usr/local/Cellar/dnscontrol/3.23.0... (5 files, 35.4MB)
```

```shell
dnscontrol write-types
```
```text
Downloading https://raw.githubusercontent.com/StackExchange/dnscontrol//types/dnscontrol.d.ts to types-dnscontrol.d.ts... done.
```

```shell
dnscontrol version
```
```text
dnscontrol -dev
```

In response to this, I opened https://github.com/StackExchange/dnscontrol/issues/1943.

Reviewer 2: Copy/pasting the content from https://github.com/StackExchange/dnscontrol/blob/v3.24.0/types/dnscontrol.d.ts does work in JetBrains IDE (PhpStorm/GoLand)

Reviewer: I just noticed that too (https://github.com/StackExchange/dnscontrol/pull/1942#issuecomment-1382423860), should hopefully be fixed for future releases by that PR though.

```bash
dnscontrol write-types
```

This file has all the information your editor or IDE needs.  It must be in the same directory as the `dnsconfig.js` file you are editing.

NOTE: Re-run the `dnscontrol write-types` command any time you upgrade

> dnscontrol. Because it is generated from the command, it will always be correct

Reviewer 2: ```suggestion
DNSControl. Because it is generated from the command, it will always be correct
```

for the version of DNSControl you are using.

2. Tell your editor

At this point some features (autocomplete) will work. However to get the full experience, including
type checking (i.e. red squiggly underlines when you misuse APIs), there is one more step.

Add this comment to the top of your `dnsconfig.js` file:

> ```

Reviewer 2: ```suggestion
```javascript
```

// @ts-check
```

That should be all you need to do!

> If your edit requires extra steps, please [file a bug](https://github.com/StackExchange/dnscontrol/issues) and we'll update this page.

Reviewer 2: ```suggestion
If your editor requires extra steps, please [file a bug](https://github.com/StackExchange/dnscontrol/issues) and we'll update this page.
```

### Bugs?

**Bugs?**  Not all features of DNSControl work perfectly at the moment. Please report bugs and feature requests on https://github.com/StackExchange/dnscontrol/issues

**This is experimental.** This feature is currently experimental. We might change the installation instructions as we find better ways to enable this.

## Known bugs

## Bug: `CLI_DEFAULTS` not implemented

Bug: Values passed to `CLI_DEFAULTS` (and the corresponding `-v` command-line option) don’t show up as global variables

Workaround: create a new `.d.ts` file in the same folder as your `dnsconfig.js` file. In that file, add the following line for each variable you want to use (replacing `VARIABLE_NAME` with the name of the variable. 

> ```

Reviewer 2: ```suggestion
```javascript
```

Reviewer: ````suggestion
```typescript
````

(`declare` isn’t a a JS thing`)

declare const VARIABLE_NAME: string;
```

This will tell TypeScript that the variable exists, and that it’s a string.

> ## Bug: `FETCH` not always accurate

Reviewer 2: ```suggestion
## Known issue: `FETCH` not always accurate
```

Bug: `FETCH` is always shown as available, even if you don’t run DNSControl with the `--allow-fetch` flag.

> Workaround: None

Reviewer 2: Based on https://github.com/StackExchange/dnscontrol/pull/1940/files#r1070092563, delete this line.

```suggestion
```

Bug: Values passed to `CLI_DEFAULTS` (and the corresponding `-v` command-line option) don’t show up as global variables

> Workaround: create a new `.d.ts` file in the same folder as your `dnsconfig.js` file. In that file, add the following line for each variable you want to use (replacing `VARIABLE_NAME` with the name of the variable. 

Reviewer: Forgot a `)`, oops!

```suggestion
Workaround: create a new `.d.ts` file in the same folder as your `dnsconfig.js` file. In that file, add the following line for each variable you want to use (replacing `VARIABLE_NAME` with the name of the variable).
```

Reviewer: Also, I don’t think literally naming the file `.d.ts` would work, I should have been more explicit here

```suggestion
Workaround: create a new file with the `.d.ts` extension in the same folder as your `dnsconfig.js` file (such as `variables.d.ts`, but the name doesn’t matter). In that file, add the following line for each variable you want to use (replacing `VARIABLE_NAME` with the name of the variable).
```
